### PR TITLE
Added local variables and general variable declaration nodes.

### LINKED ISSUE #27: Local variables
`Variable` represents a name-kind pair. A `Variable` can either come from somewhere in the program or from an external source. Right now we have `Interface` as one source. To add local variables, I propose we abstract `Interface` into `VariableSource`:

```typescript
interface VariableSource extends Hashable {
  variable: Variable;
  source(): string;
}

class Interface implements VariableSource { ... }
class Local implements VariableSource { ... }
```

### PR DESCRIPTION
**Pull request by abhimadan**
As a result, `SyntaxNode.dependencies()` is no longer necessary.

Fixes #27.

### CODE REVIEW

**Comment by abhimadan** [comment]
FYI @AndrewMcBurney I removed the check to avoid mutating `Const` interface variables, because I didn't want an interface variable-specific property to be used in a `Reference` used by local and interface variables. This is probably something we can check after building the AST.

**Comment by abhimadan** [comment]
Also, to add to my previous comment, the `const` modifier is orthogonal to interface qualifiers, so we should move it out separately and do const checks at the base `Variable` level.

File: src/variable.ts [hunk 1]
```diff
@@ -1,24 +1,24 @@
...
-            case MetaKind.Basic:
-            case MetaKind.Struct:
-                return `${this.type.name} ${this.name};`;
-            case MetaKind.Array:
-                return `${this.type.children[0].name} ${this.name}[];`;
-        }
+        return this.variableSrc.declaration();
+    }
+
+    public name(): string {
```
Comment (AndrewMcBurney) [inline src/variable.ts]: 👍 Demeter would be proud.

File: src/variablesource.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    public readonly srcType: Type;
+
+    constructor(type: Type, name: string) {
+        this.srcType = type;
+        this.srcName = name;
+    }
+
+    public declaration(): string {
+        switch (this.srcType.metakind) {
+            case MetaKind.Basic:
```
Comment (AndrewMcBurney) [inline src/variablesource.ts]: What do we return in the `MetaKind.Basic` case?

Comment (abhimadan) [reply src/variablesource.ts]: It's identical to the `MetaKind.Struct` case. It falls through to the next case in the switch statement.

Comment (AndrewMcBurney) [reply src/variablesource.ts]: Alright, sounds good - just wanted to make sure that was intended 👍

**Pull request merged**